File: CS/WindowsFormsApplication152/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
using System.Collections;
using DevExpress.XtraEditors;
using DevExpress.Data.Linq.Helpers;
using DevExpress.XtraEditors.Filtering;
using DevExpress.XtraEditors.Repository;
using DevExpress.Data.Filtering.Helpers;

namespace DXSample
{
    public partial class Form1 : XtraForm
    {
        public Form1()
        {
            InitializeComponent();
        }

        IQueryable<Products> source;
        private void OnLoad(object sender, EventArgs e)
        {
            var products = new XPQuery<Products>(unitOfWork1);
            
            source = from product in products
                          where product.Quantity >= 20
                          select product;
           
            gridControl1.DataSource = source.ToList();
            CreateFilterColumns();
        }

        private void CreateFilterColumns()
        {
            filterControl1.FilterColumns.Add(new UnboundFilterColumn("ProductID", "ProductID", typeof(int),
                new RepositoryItemSpinEdit(), FilterColumnClauseClass.Generic));
            filterControl1.FilterColumns.Add(new UnboundFilterColumn("Quantity", "Quantity", typeof(int),
                new RepositoryItemSpinEdit(), FilterColumnClauseClass.Generic));
            filterControl1.FilterColumns.Add(new UnboundFilterColumn("Price", "Price", typeof(decimal),
               new RepositoryItemSpinEdit(), FilterColumnClauseClass.Generic));
            filterControl1.FilterColumns.Add(new UnboundFilterColumn("ProductName", "ProductName", typeof(string),
               new RepositoryItemTextEdit(), FilterColumnClauseClass.String));
            filterControl1.FilterColumns.Add(new UnboundFilterColumn("Discontinued", "Discontinued", typeof(bool),
               new RepositoryItemCheckEdit(), FilterColumnClauseClass.Generic));
        }

        private void OnUpdateFilter(object sender, EventArgs e)
        {
            gridControl1.BeginUpdate();
            try
            {
                IQueryable<Products> filteredData = source.AppendWhere(filterControl1.FilterCriteria) as IQueryable<Products>;
                gridControl1.DataSource = null;
                gridControl1.DataSource = filteredData.ToList();
            }
            finally
            {
                gridControl1.EndUpdate();
            }
        }

        private void OnSaveButtonClick(object sender, EventArgs e)
        {
            try
            {
                unitOfWork1.CommitChanges();
            }
            catch (Exception exc){
                MessageBox.Show(exc.ToString());
                unitOfWork1.ReloadChangedObjects();
            }
        }
    }
}

File: VB/WindowsFormsApplication152/Form1.vb
Imports System
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.Xpo
Imports DevExpress.XtraEditors
Imports DevExpress.Data.Linq.Helpers
Imports DevExpress.XtraEditors.Filtering
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.Data.Filtering.Helpers
Imports DevExpress.Data.Linq

Namespace DXSample

    Public Partial Class Form1
        Inherits XtraForm

        Public Sub New()
            InitializeComponent()
        End Sub

        Private source As IQueryable(Of Products)

        Private Overloads Sub OnLoad(ByVal sender As Object, ByVal e As EventArgs)
            Dim products = New XPQuery(Of Products)(unitOfWork1)
            source = From product In products Where product.Quantity >= 20 Select product
            gridControl1.DataSource = source.ToList()
            CreateFilterColumns()
        End Sub

        Private Sub CreateFilterColumns()
            filterControl1.FilterColumns.Add(New UnboundFilterColumn("ProductID", "ProductID", GetType(Integer), New RepositoryItemSpinEdit(), FilterColumnClauseClass.Generic))
            filterControl1.FilterColumns.Add(New UnboundFilterColumn("Quantity", "Quantity", GetType(Integer), New RepositoryItemSpinEdit(), FilterColumnClauseClass.Generic))
            filterControl1.FilterColumns.Add(New UnboundFilterColumn("Price", "Price", GetType(Decimal), New RepositoryItemSpinEdit(), FilterColumnClauseClass.Generic))
            filterControl1.FilterColumns.Add(New UnboundFilterColumn("ProductName", "ProductName", GetType(String), New RepositoryItemTextEdit(), FilterColumnClauseClass.String))
            filterControl1.FilterColumns.Add(New UnboundFilterColumn("Discontinued", "Discontinued", GetType(Boolean), New RepositoryItemCheckEdit(), FilterColumnClauseClass.Generic))
        End Sub

        Private Sub OnUpdateFilter(ByVal sender As Object, ByVal e As EventArgs)
            gridControl1.BeginUpdate()
            Try
                Dim converter As CriteriaToExpressionConverter = New CriteriaToExpressionConverter()
                Dim filteredData As IQueryable(Of Products) = TryCast(source.AppendWhere(converter, filterControl1.FilterCriteria), IQueryable(Of Products))
                gridControl1.DataSource = Nothing
                gridControl1.DataSource = filteredData.ToList()
            Finally
                gridControl1.EndUpdate()
            End Try
        End Sub

        Private Sub OnSaveButtonClick(ByVal sender As Object, ByVal e As EventArgs)
            Try
                unitOfWork1.CommitChanges()
            Catch exc As Exception
                Call MessageBox.Show(exc.ToString())
                unitOfWork1.ReloadChangedObjects()
            End Try
        End Sub
    End Class
End Namespace
